cluster-autoscaler/core/filter_out_schedulable_test.go

```diff
@@ -109,11 +110,96 @@ func TestFilterOutSchedulableByPacking(t *testing.T) {
 
 			stillPendingPods, err := filterOutSchedulableByPacking(tt.pendingPods, clusterSnapshot, predicateChecker, 10)
 			assert.NoError(t, err)
-			assert.ElementsMatch(t, stillPendingPods, tt.expectedPendingPods)
+			assert.ElementsMatch(t, stillPendingPods, tt.expectedPendingPods, "pending pods differ")
 
 			// Check if snapshot was correctly modified
-			podsInSnapshot, _ := clusterSnapshot.GetAllPods()
-			assert.ElementsMatch(t, podsInSnapshot, tt.expectedPodsInSnapshot)
+			podsInSnapshot := clusterSnapshot.GetAllPods()
+			assert.ElementsMatch(t, podsInSnapshot, tt.expectedPodsInSnapshot, "pods in snapshot differ")
 		})
 	}
 }
+
+func BenchmarkFilterOutSchedulableByPacking(b *testing.B) {
+	tests := []struct {
+		name          string
+		nodes         int
+		scheduledPods int
+		pendingPods   int
+	}{
+		{
+			name:          "nothing",
+			nodes:         1,
+			scheduledPods: 30,
+			pendingPods:   1000,
+		},
+		{
+			name:          "small",
+			nodes:         10,
+			scheduledPods: 300,
+			pendingPods:   1000,
+		},
+		{
+			name:          "medium",
+			nodes:         100,
+			scheduledPods: 3000,
+			pendingPods:   1000,
+		},
+		{
+			name:          "large",
+			nodes:         200,
+			scheduledPods: 200,
+			pendingPods:   60000,
+		},
+		{
+			name:          "1k",
+			nodes:         1000,
+			scheduledPods: 1000,
+			pendingPods:   12000,
+		},
+	}
+	snapshots := map[string]func() simulator.ClusterSnapshot{
+		"basic": func() simulator.ClusterSnapshot { return simulator.NewBasicClusterSnapshot() },
+		"delta": func() simulator.ClusterSnapshot { return simulator.NewDeltaClusterSnapshot() },
+	}
+	for snapshotName, snapshotFactory := range snapshots {
+		for _, tc := range tests {
+			b.Run(fmt.Sprintf("%s: %d nodes %d scheduled %d pending", snapshotName, tc.nodes, tc.scheduledPods, tc.pendingPods), func(b *testing.B) {
+				pendingPods := make([]*apiv1.Pod, tc.pendingPods, tc.pendingPods)
+				for i := 0; i < tc.pendingPods; i++ {
+					pendingPods[i] = BuildTestPod(fmt.Sprintf("p-%d", i), 1000, 2000000)
+				}
+				scheduledPods := []*apiv1.Pod{}
+				nodes := make([]*apiv1.Node, tc.nodes, tc.nodes)
+				for i := 0; i < tc.nodes; i++ {
+					nodes[i] = BuildTestNode(fmt.Sprintf("n-%d", i), 2000, 200000)
+					SetNodeReadyState(nodes[i], true, time.Time{})
+				}
+
+				predicateChecker, err := simulator.NewTestPredicateChecker()
+				assert.NoError(b, err)
+
+				clusterSnapshot := snapshotFactory()
+				if err := clusterSnapshot.AddNodes(nodes); err != nil {
+					assert.NoError(b, err)
+				}
+
+				for _, pod := range scheduledPods {
+					if err := clusterSnapshot.AddPod(pod, pod.Spec.NodeName); err != nil {
+						assert.NoError(b, err)
+					}
+				}
+				b.ResetTimer()
+
+				for i := 0; i < b.N; i++ {
+					if stillPending, err := filterOutSchedulableByPacking(pendingPods, clusterSnapshot, predicateChecker, 10); err != nil {
```

Comment:
I'll rename things to clarify this, but all of the pending pods are unschedulable in this scenario (too big for the existing nodes). 

Fork() is problematic because for basic snapshot, it can dominate the cost (I've tried using b.StopTimer() and b.StartTimer(), but they seem to hang if the cost of excluded operation is too big). In real use case, we wouldn't be forking at this point, so including it in measurement doesn't make much sense.

Reply:
Makes sense if all pods are unschedulable. Maybe add a comment to make this part more obvious?

Reply:
Done